aaaaAAAA/procedural_duckies.py

```diff
@@ -0,0 +1,127 @@
+import random
+import sys
+from collections import namedtuple
+from colorsys import hls_to_rgb
+from pathlib import Path
+from typing import Optional
+
+from PIL import Image, ImageChops
+
+ProceduralDucky = namedtuple("ProceduralDucky", "image hat equipment outfit")
+DuckyColors = namedtuple("DuckyColors", "eye wing body beak")
+Color = tuple[int, int, int]
+
+DUCKY_SIZE = (499, 600)
+ASSETS_PATH = Path("assets/duck-builder")
+
+HAT_CHANCE = .7
+EQUIPMENT_CHANCE = .4
+OUTFIT_CHANCE = .5
+
+
+def make_ducky() -> ProceduralDucky:
+    """Generate a fully random ducky and returns a ProceduralDucky object."""
+    return ProceduralDuckyGenerator().generate()
+
+
+class ProceduralDuckyGenerator:
+    """Temporary class used to generate a ducky."""
+
+    templates = {
+        int(filename.name[0]): Image.open(filename) for filename in (ASSETS_PATH / "silverduck templates").iterdir()
+    }
+    hats = [
+        (filename.stem, Image.open(filename)) for filename in (ASSETS_PATH / "accessories/hats").iterdir()
+    ]
+    equipments = [
+        (filename.stem, Image.open(filename)) for filename in (ASSETS_PATH / "accessories/equipment").iterdir()
+    ]
+    outfits = [
+        (filename.stem, Image.open(filename)) for filename in (ASSETS_PATH / "accessories/outfits").iterdir()
+    ]
+
+    def __init__(self) -> None:
+        self.output: Image.Image = Image.new("RGBA", DUCKY_SIZE, color=(0, 0, 0, 0))
+        self.colors = self.make_colors()
+
+        self.hat = None
+        self.equipment = None
+        self.outfit = None
+
+    def generate(self) -> ProceduralDucky:
+        """Actually generate the ducky."""
+        self.apply_layer(self.templates[5], self.colors.beak)
+        self.apply_layer(self.templates[4], self.colors.body)
+        self.apply_layer(self.templates[3], self.colors.wing)
+        self.apply_layer(self.templates[2], self.colors.eye)
+        self.apply_layer(self.templates[1], self.colors.eye)
+
+        if random.random() < OUTFIT_CHANCE:
+            outfit_type = random.choice(self.outfits)
+            self.apply_layer(outfit_type[1])
+            self.outfit = outfit_type[0]
+
+        if random.random() < HAT_CHANCE:
+            hat_type = random.choice(self.hats)
+            self.apply_layer(hat_type[1])
+            self.hat = hat_type[0]
```

Comment:
Outfits should go above hats, since the hat `maul` covers the outfit `eyepatch`.

Reply:
On the other hand, logically hats should cover any facial outfits. Perhaps we should exempt the `maul` hat for now.

Reply:
The ninja outfit with a hat _under it_ is a lot worse than one hat covering the eyepatch, though. The former would look truly broken.

I think we will need some custom logics. For the ninja, it's better if only combined with equipment. The eyepatch should probably be a hat, not an outfit.